### PR TITLE
Replace `KtLintMultiRule`

### PR DESCRIPTION
**Pull request by marschwar**
This relates to #5192 and removes the last `MutliRule` usage in the detekt code base.

Unfortunately I had to change the documentation generation mechanism here as well as it did not work when the rule list has a function call attached such as `RuleSet("formatting", listOf(Rule1(config).sorted())`.

### CODE REVIEW

**Comment by marschwar** [comment]
Coming back to [your comment](https://github.com/detekt/detekt/issues/5192#issuecomment-1250374044) @chao2zhang 


> The reason behind this is because there are internal orders to execute ktlint rules, hence we need to extract the ordering logic into Detekt's Analyzer.

Looking at the analyzer it seems to me that the execution order of the rules within a rule set is the same as the order returned by the rule set provider. I have not tried to write a test for it though. Do you think there needs to be some more work around that?

File: detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/FormattingProvider.kt [hunk 1]
```diff
@@ -23,13 +99,94 @@ class FormattingProvider : RuleSetProvider {
...
+            TypeParameterListSpacing(config),
+        ).sortedWith(FormattingRuleComparator)
+    )
 
     companion object {
         @Configuration("if android style guides should be preferred")
         val android by ruleSetConfig(false)
-
-        @Configuration("if rules should auto correct style violation")
-        val autoCorrect by ruleSetConfig(true)
```
Comment (marschwar) [inline detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/FormattingProvider.kt:33]: This property was seemingly unused. What am I missing?

**Comment by BraisGabin** [comment]
This can be merged after 1.23.0, right? As it is a big change and it doesn't provide any feature for the user.

**Comment by marschwar** [comment]
Yes, it is just a prerequisite for the deprecation removal and for whatever we end up going with the formatting rule set.

**Review by BraisGabin — APPROVED** [review]
😍 LGTM!

**Comment by chao2zhang** [comment]
> Coming back to [your comment](https://github.com/detekt/detekt/issues/5192#issuecomment-1250374044) @chao2zhang
> 
> > The reason behind this is because there are internal orders to execute ktlint rules, hence we need to extract the ordering logic into Detekt's Analyzer.
> 
> Looking at the analyzer it seems to me that the execution order of the rules within a rule set is the same as the order returned by the rule set provider. I have not tried to write a test for it though. Do you think there needs to be some more work around that?

- The ordering logic can be found in [Analyzer](https://github.com/detekt/detekt/blob/ebbb8402c856103dc7d95b192d9f03cb034acdd8/detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/Analyzer.kt#L119) - So the order is actually determined by [FormattingProvider.kt](https://github.com/detekt/detekt/pull/6094/files/37c820c8848749c3688799a86242c03991db638e#diff-8f53a55eb4739b329546dd3dba12c13fff801f2a11f9ad97a51f8aa9bca927b3)'s autoCorrect config.

- Before this PR, all ktlint rules are supposed to run ahead of other detekt rules, simply because `is io.gitlab.arturbosch.detekt.api.MultiRule -> rule.rules.any { it.autoCorrect }`
- For maximum compatibility, we should still consider maintaining this behavior though

File: detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/FormattingProvider.kt (hunk 1 above)
Comment (chao2zhang) [reply detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/FormattingProvider.kt:33]: See https://github.com/detekt/detekt/pull/6094#issuecomment-1546740912

File: detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/FormattingRuleComparator.kt [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package io.gitlab.arturbosch.detekt.formatting
+
+/**
+ * This serves as weak heuristic to order the wrapped rules according to their visitor modifiers.
+ * Currently only RunAsLateAsPossible is supported.
+ */
+internal object FormattingRuleComparator : Comparator<FormattingRule> {
+    override fun compare(o1: FormattingRule, o2: FormattingRule): Int {
```
Comment (chao2zhang) [inline detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/FormattingRuleComparator.kt:8]: Ktlint also defines `VisitorModifier.RunAfterRule`, and we should respect that

Comment (marschwar) [reply detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/FormattingRuleComparator.kt:8]: Yap, that is described in #5259

**Comment by marschwar** [comment]
> * The ordering logic can be found in [Analyzer](https://github.com/detekt/detekt/blob/ebbb8402c856103dc7d95b192d9f03cb034acdd8/detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/Analyzer.kt#L119) - So the order is actually determined by [FormattingProvider.kt](https://github.com/detekt/detekt/pull/6094/files/37c820c8848749c3688799a86242c03991db638e#diff-8f53a55eb4739b329546dd3dba12c13fff801f2a11f9ad97a51f8aa9bca927b3)'s autoCorrect config.
> 
> * Before this PR, all ktlint rules are supposed to run ahead of other detekt rules, simply because `is io.gitlab.arturbosch.detekt.api.MultiRule -> rule.rules.any { it.autoCorrect }`
> 
> * For maximum compatibility, we should still consider maintaining this behavior though

Thank you for pointing this out. I can see why rules that might change the source code (autocorrect==true) should run first. That totally makes sense. But why must the other formatting rules run before all other rules? Honestly this feels more like a coincidence to me.

File: detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/FormattingRuleComparator.kt (hunk 2 above)
Comment (chao2zhang) [reply detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/FormattingRuleComparator.kt:8]: 😭. Would you mind if I contribute to this PR as well?

Comment (marschwar) [reply detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/FormattingRuleComparator.kt:8]: I do not mind at all. But I would not want to fix the `RunAfterRule` issue with this pr or do you see them as strongly connected?

Comment (chao2zhang) [reply detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/FormattingRuleComparator.kt:8]: Either way is fine to me - It seems that we agreed that this PR can be merged post 1.23

File: detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/FormattingProvider.kt (hunk 1 above)
Comment (chao2zhang) [reply detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/FormattingProvider.kt:33]: So this Kotlin property is not used in production
- It is used to generate the docs
- It is used to generate the config: https://github.com/detekt/detekt/blob/main/detekt-formatting/src/main/resources/config/config.yml
- Hence the actual config property "autoCorrect" was previously used

**Review by chao2zhang — APPROVED** [review]

File: detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/FormattingProvider.kt (hunk 1 above)
Comment (marschwar) [reply detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/FormattingProvider.kt]: Thank you. I have reverted the change.

**Review by cortinico — APPROVED** [review]

**Pull request merged**